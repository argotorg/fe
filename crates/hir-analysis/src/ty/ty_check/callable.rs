use hir::{
    hir_def::{kw, CallArg, ExprId, GenericArgListId},
    span::{expr::LazyCallArgListSpan, params::LazyGenericArgListSpan, DynLazySpan},
};
use if_chain::if_chain;

use super::TyChecker;
use crate::{
    ty::{
        diagnostics::{BodyDiag, FuncBodyDiag, FuncBodyDiagAccumulator},
        func_def::FuncDef,
        ty_def::{TyBase, TyData, TyId},
        ty_lower::lower_generic_arg_list,
    },
    HirAnalysisDb,
};

pub(super) struct Callable {
    func_def: FuncDef,
    generic_args: Vec<TyId>,
}

impl Callable {
    pub(super) fn new(
        db: &dyn HirAnalysisDb,
        ty: TyId,
        span: DynLazySpan,
    ) -> Result<Self, FuncBodyDiag> {
        let (base, args) = ty.decompose_ty_app(db);

        if base.is_ty_var(db) {
            return Err(BodyDiag::TypeMustBeKnown(span).into());
        }

        let TyData::TyBase(TyBase::Func(func_def)) = base.data(db) else {
            return Err(BodyDiag::not_callable(db, span, ty).into());
        };

        let params = ty.generic_args(db);
        assert_eq!(params.len(), args.len());

        Ok(Self {
            func_def: *func_def,
            generic_args: args.to_vec(),
        })
    }

    pub(super) fn ret_ty(&self, db: &dyn HirAnalysisDb) -> TyId {
        self.func_def.ret_ty(db).instantiate(db, &self.generic_args)
    }

    pub(super) fn unify_generic_args(
        &mut self,
        tc: &mut TyChecker,
        args: GenericArgListId,
        span: LazyGenericArgListSpan,
    ) -> bool {
        let db = tc.db;
        let hir_db = db.as_hir_db();

        if !args.is_given(hir_db) {
            return true;
        }

        let given_args = lower_generic_arg_list(db, args, tc.env.scope());
        let offset = self.func_def.offset_to_explicit_params_position(db);
        let current_args = &mut self.generic_args[offset..];

        if current_args.len() != given_args.len() {
            let diag = BodyDiag::CallGenericArgNumMismatch {
                primary: span.into(),
                def_span: self.func_def.name_span(db),
                given: given_args.len(),
                expected: current_args.len(),
            }
            .into();
            FuncBodyDiagAccumulator::push(db, diag);

            return false;
        }

        for (i, (&given, arg)) in given_args.iter().zip(current_args.iter_mut()).enumerate() {
            *arg = tc.equate_ty(given, *arg, span.arg(i).into());
        }

        true
    }

    pub(super) fn check_args(
        &self,
        tc: &mut TyChecker,
        args: &[CallArg],
        span: LazyCallArgListSpan,
        receiver: Option<(ExprId, TyId)>,
    ) {
        let db = tc.db;

        let expected_arity = self.func_def.arg_tys(db).len();
        let given_arity = if receiver.is_some() {
            args.len() + 1
        } else {
            args.len()
        };
        if given_arity != expected_arity {
            let diag = BodyDiag::CallArgNumMismatch {
                primary: span.into(),
                def_span: self.func_def.name_span(db),
                given: given_arity,
                expected: expected_arity,
            }
            .into();
            FuncBodyDiagAccumulator::push(db, diag);
            return;
        }

        let mut expected_args = self.func_def.arg_tys(db).iter();
        if let Some(receiver) = receiver {
            let expected = expected_args
                .next()
                .unwrap()
                .instantiate(db, &self.generic_args);
            tc.equate_ty(receiver.1, expected, receiver.0.lazy_span(tc.body()).into());
        };

        for (i, (given, expected)) in args.iter().zip(expected_args).enumerate() {
            if_chain! {
                if let Some(expected_label) = self.func_def.param_label(db, i);
                let given_label = given.label_eagerly(db.as_hir_db(), tc.body());
                if expected_label != kw::SELF;
                if Some(expected_label) != given_label;
                then {
                    let primary = if given.label.is_some() {
                        span.arg(i).label().into()
                    } else {
                        span.arg(i).into()
                    };
                    let diag = BodyDiag::CallArgLabelMismatch {
                        primary,
                        def_span: self.func_def.name_span(db),
                        given: given_label,
                        expected: expected_label,
                    }
                    .into();
                    FuncBodyDiagAccumulator::push(db, diag);
                }
            }

            let expected = expected.instantiate(db, &self.generic_args);
            let given_ty = tc.fresh_ty();
            let given_ty = tc.check_expr(given.expr, given_ty).ty;
            tc.equate_ty(given_ty, expected, span.arg(i).expr().into());
        }
    }
}
