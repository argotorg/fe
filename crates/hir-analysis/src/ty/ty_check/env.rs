use hir::{
    hir_def::{
        scope_graph::ScopeId, Body, BodyKind, Expr, ExprId, Func, IdentId, Partial, Pat, PatId,
        Stmt, StmtId,
    },
    span::DynLazySpan,
};
use rustc_hash::FxHashMap;

use super::TypedBody;
use crate::{
    ty::{
        ty_def::{InvalidCause, TyId},
        ty_lower::lower_hir_ty,
        unify::UnificationTable,
    },
    HirAnalysisDb,
};

pub(super) struct TyCheckEnv<'db> {
    db: &'db dyn HirAnalysisDb,
    body: Body,

    pat_ty: FxHashMap<PatId, TyId>,
    expr_ty: FxHashMap<ExprId, TypedExpr>,

    var_env: Vec<BlockEnv>,

    pending_vars: FxHashMap<IdentId, LocalBinding>,

    loop_stack: Vec<StmtId>,
}

impl<'db> TyCheckEnv<'db> {
    pub(super) fn new_with_func(db: &'db dyn HirAnalysisDb, func: Func) -> Result<Self, ()> {
        let hir_db = db.as_hir_db();
        let Some(body) = func.body(hir_db) else {
            return Err(());
        };

        let mut env = Self {
            db,
            body,
            pat_ty: FxHashMap::default(),
            expr_ty: FxHashMap::default(),
            var_env: vec![BlockEnv::new(func.scope(), 0)],
            pending_vars: FxHashMap::default(),
            loop_stack: Vec::new(),
        };

        env.enter_scope(body.expr(hir_db));

        let Some(params) = func.params(hir_db).to_opt() else {
            return Err(());
        };

        for (idx, param) in params.data(hir_db).iter().enumerate() {
            let Some(name) = param.name() else {
                continue;
            };

            let mut ty = match param.ty {
                Partial::Present(hir_ty) => lower_hir_ty(db, hir_ty, func.scope()),
                Partial::Absent => TyId::invalid(db, InvalidCause::Other),
            };

            if !ty.is_star_kind(db) {
                ty = TyId::invalid(db, InvalidCause::Other);
            }
            let var = LocalBinding::Param {
                idx,
                ty,
                is_mut: param.is_mut,
            };

            env.var_env.last_mut().unwrap().register_var(name, var);
        }

        Ok(env)
    }

    pub(super) fn typed_expr(&self, expr: ExprId) -> Option<TypedExpr> {
        self.expr_ty.get(&expr).copied()
    }

    pub(super) fn binding_def_span(&self, binding: LocalBinding) -> DynLazySpan {
        binding.def_span(self)
    }

    pub(super) fn binding_name(&self, binding: LocalBinding) -> IdentId {
        binding.binding_name(self)
    }

    /// Returns a function if the `body` being checked has `BodyKind::FuncBody`.
    /// If the `body` has `BodyKind::Anonymous`, returns None
    pub(super) fn func(&self) -> Option<Func> {
        match self.body.body_kind(self.db.as_hir_db()) {
            BodyKind::FuncBody => self.var_env.first()?.scope.item().try_into().ok(),
            BodyKind::Anonymous => None,
        }
    }

    pub(super) fn body(&self) -> Body {
        self.body
    }

    pub(super) fn lookup_binding_ty(&self, binding: LocalBinding) -> TyId {
        match binding {
            LocalBinding::Local { pat, .. } => self
                .pat_ty
                .get(&pat)
                .copied()
                .unwrap_or_else(|| TyId::invalid(self.db, InvalidCause::Other)),

            LocalBinding::Param { ty, .. } => ty,
        }
    }

    pub(super) fn enter_scope(&mut self, block: ExprId) {
        let new_scope = match block.data(self.db.as_hir_db(), self.body) {
            Partial::Present(Expr::Block(_)) => ScopeId::Block(self.body, block),
            _ => self.scope(),
        };

        let var_env = BlockEnv::new(new_scope, self.var_env.len());
        self.var_env.push(var_env);
    }

    pub(super) fn leave_scope(&mut self) {
        self.var_env.pop().unwrap();
    }

    pub(super) fn enter_loop(&mut self, stmt: StmtId) {
        self.loop_stack.push(stmt);
    }

    pub(super) fn leave_loop(&mut self) {
        self.loop_stack.pop();
    }

    pub(super) fn current_loop(&self) -> Option<StmtId> {
        self.loop_stack.last().copied()
    }

    pub(super) fn type_expr(&mut self, expr: ExprId, typed: TypedExpr) {
        self.expr_ty.insert(expr, typed);
    }

    pub(super) fn type_pat(&mut self, pat: PatId, ty: TyId) {
        self.pat_ty.insert(pat, ty);
    }

    /// Register a pending binding which will be added when `flush_pending_vars`
    /// is called.
    pub(super) fn register_pending_binding(&mut self, name: IdentId, binding: LocalBinding) {
        self.pending_vars.insert(name, binding);
    }

    /// Flush pending bindings to the current scope environment.
    pub(super) fn flush_pending_bindings(&mut self) {
        let var_env = self.var_env.last_mut().unwrap();
        for (name, binding) in self.pending_vars.drain() {
            var_env.register_var(name, binding);
        }
    }

    pub(super) fn finish(mut self, table: &mut UnificationTable) -> TypedBody {
        self.expr_ty
            .values_mut()
            .for_each(|ty| *ty = ty.apply_subst(self.db, table));

        self.pat_ty
            .values_mut()
            .for_each(|ty| *ty = ty.apply_subst(self.db, table));

        TypedBody {
            body: Some(self.body),
            pat_ty: self.pat_ty,
            expr_ty: self.expr_ty,
        }
    }

    pub(super) fn expr_data(&self, expr: ExprId) -> &'db Partial<Expr> {
        expr.data(self.db.as_hir_db(), self.body)
    }

    pub(super) fn stmt_data(&self, stmt: StmtId) -> &'db Partial<Stmt> {
        stmt.data(self.db.as_hir_db(), self.body)
    }

    pub(super) fn scope(&self) -> ScopeId {
        self.var_env.last().unwrap().scope
    }

    pub(super) fn current_block_idx(&self) -> usize {
        self.var_env.last().unwrap().idx
    }

    pub(super) fn get_block(&self, idx: usize) -> &BlockEnv {
        &self.var_env[idx]
    }
}

pub(super) struct BlockEnv {
    pub(super) scope: ScopeId,
    pub(super) vars: FxHashMap<IdentId, LocalBinding>,
    idx: usize,
}

impl BlockEnv {
    pub(super) fn lookup_var(&self, var: IdentId) -> Option<LocalBinding> {
        self.vars.get(&var).copied()
    }

    fn new(scope: ScopeId, idx: usize) -> Self {
        Self {
            scope,
            vars: FxHashMap::default(),
            idx,
        }
    }

    fn register_var(&mut self, name: IdentId, var: LocalBinding) {
        self.vars.insert(name, var);
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub(super) struct TypedExpr {
    ty: TyId,
    is_mut: bool,
    binding: Option<LocalBinding>,
}

impl TypedExpr {
    pub(super) fn new(ty: TyId, is_mut: bool) -> Self {
        Self {
            ty,
            is_mut,
            binding: None,
        }
    }

    pub(super) fn new_binding_ref(ty: TyId, is_mut: bool, binding: LocalBinding) -> Self {
        Self {
            ty,
            is_mut,
            binding: Some(binding),
        }
    }

    pub(super) fn binding(&self) -> Option<LocalBinding> {
        self.binding
    }

    pub(super) fn swap_ty(&mut self, ty: TyId) -> TyId {
        std::mem::replace(&mut self.ty, ty)
    }

    pub(super) fn invalid(db: &dyn HirAnalysisDb) -> Self {
        Self {
            ty: TyId::invalid(db, InvalidCause::Other),
            is_mut: true,
            binding: None,
        }
    }

    pub(super) fn ty(&self) -> TyId {
        self.ty
    }

    pub(super) fn is_mutable(&self, _db: &dyn HirAnalysisDb) -> bool {
        // TODO: We need to check both type mutability and expr mutability when we have
        // `&mut` type.
        self.is_mut
    }

    pub(super) fn apply_subst(self, db: &dyn HirAnalysisDb, table: &mut UnificationTable) -> Self {
        Self {
            ty: self.ty.apply_subst(db, table),
            is_mut: self.is_mut,
            binding: self.binding,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub(super) enum LocalBinding {
    Local { pat: PatId, is_mut: bool },
    Param { idx: usize, ty: TyId, is_mut: bool },
}

impl LocalBinding {
    pub(super) fn local(pat: PatId, is_mut: bool) -> Self {
        Self::Local { pat, is_mut }
    }

    pub(super) fn is_mut(&self) -> bool {
        match self {
            LocalBinding::Local { is_mut, .. } | LocalBinding::Param { is_mut, .. } => *is_mut,
        }
    }

    pub(super) fn binding_name(&self, env: &TyCheckEnv<'_>) -> IdentId {
        let hir_db = env.db.as_hir_db();
        match self {
            Self::Local { pat, .. } => {
                let Partial::Present(Pat::Path(Partial::Present(path), ..)) =
                    pat.data(hir_db, env.body())
                else {
                    unreachable!();
                };
                *path.last_segment(hir_db).unwrap()
            }

            Self::Param { idx, .. } => {
                let func = env.func().unwrap();
                let Partial::Present(func_params) = func.params(hir_db) else {
                    unreachable!();
                };

                func_params.data(hir_db)[*idx].name().unwrap()
            }
        }
    }

    fn def_span(&self, env: &TyCheckEnv<'_>) -> DynLazySpan {
        match self {
            LocalBinding::Local { pat, .. } => pat.lazy_span(env.body).into(),
            LocalBinding::Param { idx, .. } => {
                let func = env.func().unwrap();
                func.lazy_span()
                    .params_moved()
                    .param(*idx)
                    .name_moved()
                    .into()
            }
        }
    }
}
