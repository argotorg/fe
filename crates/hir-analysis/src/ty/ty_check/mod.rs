mod env;
mod expr;
mod pat;
mod stmt;

use env::TyCheckEnv;
use hir::{
    hir_def::{
        Body, Enum, ExprId, Func, LitKind, PatId, PathId, TypeId as HirTyId,
        VariantKind as HirVariantKind,
    },
    span::DynLazySpan,
};
use rustc_hash::FxHashMap;

use super::{
    diagnostics::{BodyDiag, FuncBodyDiagAccumulator},
    ty_def::{AdtDef, AdtRef, AdtRefId, InvalidCause, Kind, TyId, TyVarUniverse},
    ty_lower::{lower_adt, lower_hir_ty},
    unify::{UnificationError, UnificationTable},
};
use crate::HirAnalysisDb;

#[salsa::tracked(return_ref)]
pub fn check_func_body(db: &dyn HirAnalysisDb, func: Func) -> TypedBody {
    let Ok(mut checker) = TyChecker::new_with_func(db, func) else {
        return TypedBody::empty();
    };

    checker.run();
    checker.finish()
}

pub struct TyChecker<'db> {
    db: &'db dyn HirAnalysisDb,
    env: TyCheckEnv<'db>,
    table: UnificationTable<'db>,
    expected: TyId,
}

impl<'db> TyChecker<'db> {
    fn new_with_func(db: &'db dyn HirAnalysisDb, func: Func) -> Result<Self, ()> {
        let env = TyCheckEnv::new_with_func(db, func)?;
        let expected_ty = match func.ret_ty(db.as_hir_db()) {
            Some(hir_ty) => {
                let ty = lower_hir_ty(db, hir_ty, func.scope());
                if ty.is_star_kind(db) {
                    ty
                } else {
                    TyId::invalid(db, InvalidCause::Other)
                }
            }
            None => TyId::unit(db),
        };

        Ok(Self::new(db, env, expected_ty))
    }

    fn run(&mut self) {
        let root_expr = self.env.body().expr(self.db.as_hir_db());
        self.check_expr(root_expr, self.expected);
    }

    fn finish(mut self) -> TypedBody {
        // TODO: check for untyped expressions and patterns.
        self.env.finish(&mut self.table)
    }

    fn new(db: &'db dyn HirAnalysisDb, env: TyCheckEnv<'db>, expected: TyId) -> Self {
        let table = UnificationTable::new(db);
        Self {
            db,
            env,
            table,
            expected,
        }
    }

    fn body(&self) -> Body {
        self.env.body()
    }

    fn lit_ty(&mut self, lit: &LitKind) -> TyId {
        match lit {
            LitKind::Bool(_) => TyId::bool(self.db),
            LitKind::Int(_) => self.table.new_var(TyVarUniverse::Integral, &Kind::Star),
            LitKind::String(s) => {
                let len_bytes = s.len_bytes(self.db.as_hir_db());
                self.table
                    .new_var(TyVarUniverse::String(len_bytes), &Kind::Star)
            }
        }
    }

    fn lower_ty(&self, hir_ty: HirTyId, span: DynLazySpan) -> TyId {
        let ty = lower_hir_ty(self.db, hir_ty, self.env.scope());
        if let Some(diag) = ty.emit_diag(self.db, span) {
            FuncBodyDiagAccumulator::push(self.db, diag.into());
        }

        ty
    }

    /// Returns the fresh type variable for pattern and expr type checking. The
    /// kind of the type variable is `*`, and the universe is `General`.
    fn fresh_ty(&mut self) -> TyId {
        self.table.new_var(TyVarUniverse::General, &Kind::Star)
    }

    fn fresh_tys_n(&mut self, n: usize) -> Vec<TyId> {
        (0..n).map(|_| self.fresh_ty()).collect()
    }

    fn unify_ty<T>(&mut self, t: T, actual: TyId, expected: TyId) -> TyId
    where
        T: Into<Typeable>,
    {
        let t = t.into();

        let actual = match self.table.unify(expected, actual) {
            Ok(()) => actual,

            Err(UnificationError::TypeMismatch) => {
                let actual = actual.apply_subst(self.db, &mut self.table);
                let expected = expected.apply_subst(self.db, &mut self.table);
                FuncBodyDiagAccumulator::push(
                    self.db,
                    BodyDiag::type_mismatch(
                        self.db,
                        t.lazy_span(self.env.body()),
                        expected,
                        actual,
                    )
                    .into(),
                );
                TyId::invalid(self.db, InvalidCause::Other)
            }

            Err(UnificationError::OccursCheckFailed) => {
                FuncBodyDiagAccumulator::push(
                    self.db,
                    BodyDiag::InfiniteOccurrence(t.lazy_span(self.env.body())).into(),
                );

                TyId::invalid(self.db, InvalidCause::Other)
            }
        };

        match t {
            Typeable::Expr(expr) => self.env.type_expr(expr, actual),
            Typeable::Pat(pat) => self.env.type_pat(pat, actual),
        }

        actual
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TypedBody {
    body: Option<Body>,
    pat_ty: FxHashMap<PatId, TyId>,
    expr_ty: FxHashMap<ExprId, TyId>,
}

impl TypedBody {
    pub fn expr_ty(&self, db: &dyn HirAnalysisDb, expr: ExprId) -> TyId {
        self.expr_ty
            .get(&expr)
            .copied()
            .unwrap_or_else(|| TyId::invalid(db, InvalidCause::Other))
    }
    pub fn pat_ty(&self, db: &dyn HirAnalysisDb, pat: PatId) -> TyId {
        self.pat_ty
            .get(&pat)
            .copied()
            .unwrap_or_else(|| TyId::invalid(db, InvalidCause::Other))
    }

    fn empty() -> Self {
        Self {
            body: None,
            pat_ty: FxHashMap::default(),
            expr_ty: FxHashMap::default(),
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq, derive_more::From)]
enum Typeable {
    Expr(ExprId),
    Pat(PatId),
}

impl Typeable {
    fn lazy_span(self, body: Body) -> DynLazySpan {
        match self {
            Self::Expr(expr) => expr.lazy_span(body).into(),
            Self::Pat(pat) => pat.lazy_span(body).into(),
        }
    }
}

#[derive(Clone, Debug, Hash, PartialEq, Eq)]
pub(crate) enum ResolvedPathData {
    Adt(AdtDef, PathId),
    Variant(Enum, usize, PathId),
}

impl ResolvedPathData {
    pub(crate) fn adt_ref(&self, db: &dyn HirAnalysisDb) -> AdtRef {
        match self {
            Self::Adt(adt, _) => adt.adt_ref(db).data(db),
            Self::Variant(enum_, _, _) => AdtRef::Enum(*enum_),
        }
    }

    pub(crate) fn data_kind(&self, db: &dyn HirAnalysisDb) -> &'static str {
        match self {
            Self::Adt(adt, _) => adt.adt_ref(db).kind_name(db),
            Self::Variant(enum_, idx, _) => {
                let hir_db = db.as_hir_db();
                match enum_.variants(hir_db).data(hir_db)[*idx].kind {
                    hir::hir_def::VariantKind::Unit => "unit variant",
                    HirVariantKind::Tuple(_) => "tuple variant",
                    HirVariantKind::Record(_) => "record variant",
                }
            }
        }
    }

    pub(crate) fn initializer_hint(&self, db: &dyn HirAnalysisDb) -> Option<String> {
        let hir_db = db.as_hir_db();

        let expected_sub_pat = match self {
            Self::Adt(_, _) => {
                let AdtRef::Struct(s) = self.adt_ref(db) else {
                    return None;
                };

                s.format_initializer_args(hir_db)
            }

            Self::Variant(enum_, idx, _) => {
                enum_.variants(hir_db).data(hir_db)[*idx].format_initializer_args(hir_db)
            }
        };

        let path = self.path().pretty_print(hir_db);
        Some(format!("{}{}", path, expected_sub_pat))
    }

    fn path(&self) -> PathId {
        match self {
            Self::Adt(_, path) => *path,
            Self::Variant(_, _, path) => *path,
        }
    }

    fn ty(&self, db: &dyn HirAnalysisDb, table: &mut UnificationTable) -> TyId {
        let adt = match self {
            Self::Adt(adt, _) => *adt,

            Self::Variant(enum_, ..) => lower_adt(db, AdtRefId::from_enum(db, *enum_)),
        };

        let adt_ty = TyId::adt(db, adt);
        adt.params(db).iter().fold(adt_ty, |ty, param| {
            let param_ty = table.new_var_from_param(*param);
            TyId::app(db, ty, param_ty)
        })
    }

    fn is_unit_variant(&self, db: &dyn HirAnalysisDb) -> bool {
        match self {
            Self::Adt(_, _) => false,
            Self::Variant(enum_, idx, _) => {
                let hir_db = db.as_hir_db();
                matches!(
                    enum_.variants(hir_db).data(hir_db)[*idx].kind,
                    hir::hir_def::VariantKind::Unit
                )
            }
        }
    }

    fn is_tuple_variant(&self, db: &dyn HirAnalysisDb) -> bool {
        match self {
            Self::Adt(_, _) => false,
            Self::Variant(enum_, idx, _) => {
                let hir_db = db.as_hir_db();
                matches!(
                    enum_.variants(hir_db).data(hir_db)[*idx].kind,
                    HirVariantKind::Tuple(_)
                )
            }
        }
    }

    fn field_tys(&self, db: &dyn HirAnalysisDb) -> Vec<TyId> {
        let (adt, idx) = match self {
            Self::Adt(adt, _) => {
                if matches!(
                    adt.adt_ref(db).data(db),
                    AdtRef::Struct(_) | AdtRef::Contract(_)
                ) {
                    (*adt, 0)
                } else {
                    return vec![];
                }
            }
            Self::Variant(enum_, idx, _) => {
                let adt = lower_adt(db, AdtRefId::from_enum(db, *enum_));
                (adt, *idx)
            }
        };

        adt.fields(db)[idx].iter_types(db).collect()
    }
}
