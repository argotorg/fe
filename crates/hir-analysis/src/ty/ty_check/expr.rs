use hir::hir_def::{Expr, ExprId, Partial};

use super::path::ResolvedPathInExpr;
use crate::ty::{
    diagnostics::{BodyDiag, FuncBodyDiagAccumulator},
    ty_check::{path::resolve_path_in_expr, TyChecker},
    ty_def::{InvalidCause, TyId},
};

impl<'db> TyChecker<'db> {
    pub(super) fn check_expr(&mut self, expr: ExprId, expected: TyId) -> TyId {
        let Partial::Present(expr_data) = self.env.expr_data(expr) else {
            let ty = TyId::invalid(self.db, InvalidCause::Other);
            self.env.type_expr(expr, ty);
            return ty;
        };

        let actual = match expr_data {
            Expr::Lit(lit) => self.lit_ty(lit),
            Expr::Block(..) => self.check_block(expr, expr_data, expected),

            Expr::Bin(..) => todo!(),
            Expr::Un(..) => todo!(),
            Expr::Call(..) => todo!(),
            Expr::MethodCall(..) => todo!(),
            Expr::Path(..) => self.check_path(expr, expr_data, expected),
            Expr::RecordInit(..) => todo!(),
            Expr::Field(..) => todo!(),
            Expr::Tuple(..) => todo!(),
            Expr::Index(..) => todo!(),
            Expr::Array(..) => todo!(),
            Expr::ArrayRep(..) => todo!(),
            Expr::If(..) => self.check_if(expr, expr_data, expected),
            Expr::Match(..) => todo!(),
            Expr::Assign(..) => todo!(),
            Expr::AugAssign(..) => todo!(),
        };

        self.unify_ty(expr, actual, expected)
    }

    fn check_block(&mut self, expr: ExprId, expr_data: &Expr, expected: TyId) -> TyId {
        let Expr::Block(stmts) = expr_data else {
            unreachable!()
        };

        if stmts.is_empty() {
            TyId::unit(self.db)
        } else {
            self.env.enter_scope(expr);
            for &stmt in stmts[..stmts.len() - 1].iter() {
                let ty = self.fresh_ty();
                self.check_stmt(stmt, ty);
            }

            let last_stmt = stmts[stmts.len() - 1];
            let res = self.check_stmt(last_stmt, expected);
            self.env.leave_scope();
            res
        }
    }

    fn check_path(&mut self, expr: ExprId, expr_data: &Expr, _expected: TyId) -> TyId {
        let path = match expr_data {
            Expr::Path(path) => path,
            _ => unreachable!(),
        };

        let Partial::Present(path) = path else {
            return TyId::invalid(self.db, InvalidCause::Other);
        };

        match resolve_path_in_expr(self, *path, expr) {
            ResolvedPathInExpr::Data(data) => {
                if data.is_unit_variant(self.db) {
                    data.ty(self.db)
                } else {
                    let diag = BodyDiag::unit_variant_expected(
                        self.db,
                        expr.lazy_span(self.body()).into(),
                        data,
                    )
                    .into();
                    FuncBodyDiagAccumulator::push(self.db, diag);

                    TyId::invalid(self.db, InvalidCause::Other)
                }
            }

            ResolvedPathInExpr::Func(func) => func.ty(self.db),

            ResolvedPathInExpr::Binding(_, binding) => self.env.lookup_binding_ty(binding),

            ResolvedPathInExpr::Diag(diag) => {
                FuncBodyDiagAccumulator::push(self.db, diag);
                TyId::invalid(self.db, InvalidCause::Other)
            }

            ResolvedPathInExpr::Invalid => TyId::invalid(self.db, InvalidCause::Other),
        }
    }

    fn check_if(&mut self, _expr: ExprId, expr_data: &Expr, expected: TyId) -> TyId {
        let Expr::If(cond, then, else_) = expr_data else {
            unreachable!()
        };

        self.check_expr(*cond, TyId::bool(self.db));

        match else_ {
            Some(else_) => {
                self.check_expr_in_new_scope(*then, expected);
                self.check_expr_in_new_scope(*else_, expected)
            }

            None => {
                // If there is no else branch, the if expression itself typed as `()`
                let expected = self.fresh_ty();
                self.check_expr_in_new_scope(*then, expected);
                TyId::unit(self.db)
            }
        }
    }

    fn check_expr_in_new_scope(&mut self, expr: ExprId, expected: TyId) -> TyId {
        self.env.enter_scope(expr);
        let ty = self.check_expr(expr, expected);
        self.env.leave_scope();

        ty
    }
}
