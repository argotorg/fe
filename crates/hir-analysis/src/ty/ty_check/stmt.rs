use hir::hir_def::{Partial, Stmt, StmtId};

use super::TyChecker;
use crate::ty::ty_def::{InvalidCause, TyId};

impl<'db> TyChecker<'db> {
    pub(super) fn check_stmt(&mut self, stmt: StmtId, expected: TyId) -> TyId {
        let Partial::Present(stmt_data) = self.env.stmt_data(stmt) else {
            return TyId::invalid(self.db, InvalidCause::Other);
        };

        match stmt_data {
            Stmt::Let(..) => self.check_let(stmt, stmt_data),

            Stmt::For(..) => todo!(),

            Stmt::While(..) => todo!(),

            Stmt::Continue => todo!(),

            Stmt::Break => todo!(),

            Stmt::Return(..) => todo!(),

            Stmt::Expr(expr) => self.check_expr_ty(*expr, expected),
        }
    }

    fn check_let(&mut self, stmt: StmtId, stmt_data: &Stmt) -> TyId {
        let Stmt::Let(pat, ascription, expr) = stmt_data else {
            unreachable!()
        };

        let span = stmt.lazy_span(self.env.body()).into_let_stmt();

        let ascription = match ascription {
            Some(ty) => self.lower_ty(*ty, span.ty_moved().into()),
            None => self.fresh_ty(),
        };

        self.check_pat(*pat, ascription);
        if let Some(expr) = expr {
            self.check_expr_ty(*expr, ascription);
        }

        self.env.flush_pending_bindings();
        TyId::unit(self.db)
    }
}
