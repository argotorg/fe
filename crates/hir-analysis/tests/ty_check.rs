mod test_db;
use std::path::Path;

use dir_test::{dir_test, Fixture};
use fe_compiler_test_utils::snap_test;
use fe_hir_analysis::ty::ty_check::check_func_body;
use test_db::HirAnalysisTestDb;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/ty_check",
    glob: "*.fe"
)]
fn test_standalone(fixture: Fixture<&str>) {
    let mut db = HirAnalysisTestDb::default();
    let path = Path::new(fixture.path());
    let file_name = path.file_name().and_then(|file| file.to_str()).unwrap();
    let (top_mod, mut prop_formatter) = db.new_stand_alone(file_name, fixture.content());

    for &func in top_mod.all_funcs(&db) {
        let Some(body) = func.body(&db) else {
            continue;
        };

        let typed_body = check_func_body(&db, func);
        for expr in body.exprs(&db).keys() {
            let ty = typed_body.expr_ty(&db, expr);
            prop_formatter.push_prop(
                func.top_mod(&db),
                expr.lazy_span(body).into(),
                ty.pretty_print(&db).to_string(),
            );
        }
    }

    let res = prop_formatter.finish(&db);
    snap_test!(res, fixture.path());
}
