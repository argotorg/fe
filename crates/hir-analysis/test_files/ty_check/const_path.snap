---
source: crates/hir-analysis/tests/ty_check.rs
expression: res
input_file: test_files/ty_check/const_path.fe
---
note: 
  ┌─ const_path.fe:6:27
  │  
6 │       fn len(self) -> usize {
  │ ╭───────────────────────────^
7 │ │         N
8 │ │     }
  │ ╰─────^ usize

note: 
  ┌─ const_path.fe:7:9
  │
7 │         N
  │         ^ usize

note: 
   ┌─ const_path.fe:17:76
   │  
17 │   fn f<const N: usize>(b5: B5, arr: [u8; N], a: A<N>, s: String<N>) -> usize {
   │ ╭────────────────────────────────────────────────────────────────────────────^
18 │ │     a.len()
19 │ │ }
   │ ╰─^ usize

note: 
   ┌─ const_path.fe:18:5
   │
18 │     a.len()
   │     ^ A<const N: usize>

note: 
   ┌─ const_path.fe:18:5
   │
18 │     a.len()
   │     ^^^^^^^ usize

note: 
   ┌─ const_path.fe:22:29
   │  
22 │   fn g(arr: [u8; SIZE]) -> u8 {
   │ ╭─────────────────────────────^
23 │ │     let s: String<SIZE> = "0123456789"
24 │ │     arr[SIZE-1]
25 │ │ }
   │ ╰─^ u8

note: 
   ┌─ const_path.fe:23:9
   │
23 │     let s: String<SIZE> = "0123456789"
   │         ^ String<10>

note: 
   ┌─ const_path.fe:23:27
   │
23 │     let s: String<SIZE> = "0123456789"
   │                           ^^^^^^^^^^^^ String<10>

note: 
   ┌─ const_path.fe:24:5
   │
24 │     arr[SIZE-1]
   │     ^^^ [u8; 10]

note: 
   ┌─ const_path.fe:24:5
   │
24 │     arr[SIZE-1]
   │     ^^^^^^^^^^^ u8

note: 
   ┌─ const_path.fe:24:9
   │
24 │     arr[SIZE-1]
   │         ^^^^ usize

note: 
   ┌─ const_path.fe:24:9
   │
24 │     arr[SIZE-1]
   │         ^^^^^^ usize

note: 
   ┌─ const_path.fe:24:14
   │
24 │     arr[SIZE-1]
   │              ^ usize
